p-ata/src/recover.rs

```diff
@@ -0,0 +1,367 @@
+        // Multisig case: must be token-program owned
+        if !recover_accounts
+            .wallet
+            .is_owned_by(recover_accounts.token_program.key())
+        {
+            log!(
+                "Error: Multisig wallet {} is not owned by token program {}",
+                recover_accounts.wallet.key(),
+                recover_accounts.token_program.key()
+            );
+            return Err(ProgramError::MissingRequiredSignature);
+        }
+
+        let wallet_data_slice = unsafe { recover_accounts.wallet.borrow_data_unchecked() };
+        let multisig_state: &Multisig = unsafe {
+            spl_token_interface::state::load::<Multisig>(wallet_data_slice)
+                .map_err(|_| ProgramError::InvalidAccountData)?
+        };
+
+        let signer_infos = &accounts[7..];
+
+        let mut num_signers = 0;
+        let mut matched = [false; MAX_SIGNERS as usize];
+
+        for signer in signer_infos.iter() {
+            for (position, key) in multisig_state.signers[0..multisig_state.n as usize]
+                .iter()
+                .enumerate()
+            {
+                if key == signer.key() && !matched[position] {
+                    if !signer.is_signer() {
+                        log!(
+                            "Error: Multisig member account {} is not a signer",
+                            signer.key()
+                        );
+                        return Err(ProgramError::MissingRequiredSignature);
+                    }
+                    matched[position] = true;
+                    num_signers += 1;
+                }
+            }
+        }
+
+        if num_signers < multisig_state.m {
+            log!(
+                "Error: Insufficient multisig signatures. Required: {}, Found: {}",
+                multisig_state.m,
+                num_signers
+            );
+            return Err(ProgramError::MissingRequiredSignature);
+        }
+    }
+
+    let amount_to_recover =
+        load_token_account(recover_accounts.nested_associated_token_account)?.amount();
+
+    let nested_mint_decimals = get_decimals_from_mint(recover_accounts.nested_mint)?;
+
+    let transfer_data = build_transfer_checked_data(amount_to_recover, nested_mint_decimals);
+
+    let transfer_metas = &[
+        AccountMeta {
+            pubkey: recover_accounts.nested_associated_token_account.key(),
+            is_writable: true,
+            is_signer: false,
+        },
+        AccountMeta {
+            pubkey: recover_accounts.nested_mint.key(),
+            is_writable: false,
+            is_signer: false,
+        },
+        AccountMeta {
+            pubkey: recover_accounts.destination_associated_token_account.key(),
+            is_writable: true,
+            is_signer: false,
+        },
+        AccountMeta {
+            pubkey: recover_accounts.owner_associated_token_account.key(),
+            is_writable: false,
+            is_signer: true,
+        },
```

Comment:
Can use the constructors:
```suggestion
        AccountMeta::writable(
                recover_accounts.nested_associated_token_account.key()
        ),
        AccountMeta::readonly(
                recover_accounts.nested_mint.key()
        ),
        AccountMeta::writable(
                recover_accounts.destination_associated_token_account.key()
        ),
        AccountMeta::readonly_signer(
                recover_accounts.owner_associated_token_account.key()
        ),
```